#' MEA RNA 2ndary structures of mm10 exons predicted by RNAfold
#'
#' The full length transcripts are extracted from mm10;
#' the transcript annotation file was downloaded from the refSeq archive-2015-07-17-14-33-26.
#' The temperature of RNAfold prediction was set at 70 degree, and
#' the gammar for MEA structure was set at 0.1. The maximum paring distance was set at 150bp.
#' 
#' For the transcripts longer than 8000bp, the predictions were conducted in windows of 2000bp with steps of 1000bp.
#' 
#' 
#'
#' @format A GRangesList object of length 34597:
#' \describe{
#'   Each element of the GRangesList represents a transcript, exept that the first element refers to the mitochondria chromosome.
#'   The GRanges within each of the GRangesList element represent the genomic locations of the MEA RNA 2ndary structures predicted by the method above.
#' }
#' @usage 
#' rBS_gr = GRanges(seqnames = rBS_df$`#SeqID`, strand = rBS_df$refStrand, ranges = IRanges(start = rBS_df$refPos, width = 1))
#' rBS_gr_filtered <- rBS_gr[!rBS_gr \%over\% Struc_mm10]
#' @note For other filtering strategies, please check \code{\link{rBS_2ndStructure_Filter}}
"Struc_mm10"

#' MEA RNA 2ndary structures of hg19 exons predicted by RNAfold
#'
#' The full length transcripts are extracted from hg19;
#' the transcript annotation file was downloaded from the refSeq archive-2015-07-17-14-32-32.
#' The temperature of RNAfold prediction was set at 70 degree, and
#' the gammar for MEA structure was set at 0.1. The maximum paring distance was set at 150bp.
#' 
#' For the transcripts longer than 8000bp, the predictions were conducted in windows of 2000bp with steps of 1000bp.
#'
#' @format A GRangesList object of length 53350:
#' \describe{
#'   Each element of the GRangesList represents a transcript, exept that the first element refers to the mitochondria chromosome.
#'   The GRanges within each of the GRangesList element represent the genomic locations of the MEA RNA 2ndary structures predicted by the method above.
#' }
#' @usage 
#' rBS_gr = GRanges(seqnames = rBS_df$`#SeqID`, strand = rBS_df$refStrand, ranges = IRanges(start = rBS_df$refPos, width = 1))
#' rBS_gr_filtered <- rBS_gr[!rBS_gr \%over\% Struc_hg19]
#' @note For other filtering strategies, please check \code{\link{rBS_2ndStructure_Filter}}
"Struc_hg19"

#' An example table of RNA bisulfite sequencing data processed by MeRanTK.
#'
#' The data comes from a subset of MeRanTK processed RNA bisulfite sample of human MCF10A cell.
#' The table is directly imported from the txt file generated by the MeRanCall of MeRanTK with minimum read coverage of 10 and maximum conversion rate of 1.  
#' 
#' @format A data.frame object with 100000 rows and 20 columns:
#' \describe{
#'   #SeqID : the chromosomal IDs provided by the reference trasncripts of hg19.
#'   refStrand : the strands of the sites.
#'   refPos: the location of the sites.
#'   p-value_mState: the hypergeometric p value related to the statistical significance of the methylation state.
#'  ...
#'  }
#' @source \url{http://icbi.at/software/meRanTK/}
"rBS_df"

#' An GRangesList created by exonsBy(txdb).
#'
#' The txdb is imported from the gtf file downloaded from the mm10 refSeq archive-2015-07-17-14-33-26.
#'  
#' @format A GRangesList object with length 35119.
"exbytx_mm10"

#' An GRangesList created by exonsBy(txdb).
#'
#' The txdb is imported from the gtf file downloaded from the hg19 refSeq archive-2015-07-17-14-32-32.
#'  
#' @format A GRangesList object with length 54070.
"exbytx_hg19"

